Added support for BosonOperator and BoseHubbard models

josh146 opened the pull request:
This pull request addresses issue #151, and begins adding bosonic support to OpenFermion.

I've tried to keep the changes minimal, and just introduce the base structure of bosonic operators. This has led to some weird naming conventions, such as `normal_ordered_boson` and `normal_ordered_quad`, so as not to effect the already existing `normal_ordered` function which acts on fermion operators - this should probably be modified at some point to bring them into alignment.

Among the changes in this pull request:

* `BosonOperator` has been added, which allows the creation of bosonic ladder operators satisfying the bosonic commutation relations. In addition, it supports the methods `is_boson_preserving` and `is_normal_ordered`, to bring it inline with the `FermionOperator`.

* `QuadOperator` has been added, which allows the creation of the canonical position and momentum quadrature operators. This supports the method `is_gaussian`, which returns True if the resulting operator transforms the bosonic systems via a Gaussian symplectic transformation. In addition, `normal_ordered_quad` uses the canonical commutation relation [q,p]=i hbar to rewrite the operator such that all instances of q appear to the left of p.

* `get_boson_operator` and `get_quad_operator` conversions have been added, to allow the user to transform between these two representations.

* `weyl_polynomial_quantization` allows the user to use Weyl quantization to find the quadrature operator representation of a polynomial in the bosonic phase space, using McCoy's formula.

* `symmetric_ordering` is a transformation that returns the symmetric ordering of a `BosonOperator` or `QuadOperator`, important if working with Wigner pseudo-probability distributions.

* Sparse operator methods have been added, for easily converting the `BosonOperator` or `QuadOperator` into sparse matrix form.

* The `BoseHubbard` model has been added to Hamiltonians, allowing the user to quickly generate Bose-Hubbard systems with optional on-site or dipole interactions.

* A Jupyter notebook tutorial for bosonic systems (based on the fermionic notebook) has been added to the examples folder.

Tests have been added for each of these new objects, and it has been linted against PEP8. This is a preliminary pull request - once a structure exists for bosonic Hamiltonians, we can start to look at methods for constructing qubit representations (see https://arxiv.org/abs/1802.07347).

babbush commented:
Thank you for this excellent looking contribution! Due the unusually large size of this pull request, it may take us a few days to review. But don't worry, we're on it!

babbush commented on src/openfermion/transforms/_weyl_ordering_test.py:
```diff
@@ -0,0 +1,202 @@
+import numpy
+
+from openfermion.ops import (FermionOperator, BosonOperator,
+                             QuadOperator,
```
alphabetical imports

babbush commented on src/openfermion/utils/_special_operators.py:
```diff
@@ -309,3 +309,23 @@ def number_operator(n_orbitals, orbital=None, coefficient=1.):
     return operator
+
+
+def boson_number_operator(n_modes, mode=None, coefficient=1.):
```
I am thinking perhaps we should have one function with a switch between bosons and fermions. Just a thought.

josh146 replied:
I agree - should I merge them into a single number_operator?

babbush replied:
sure, thanks!

josh146 commented on src/openfermion/hamiltonians/_bose_hubbard.py:
```diff
@@ -0,0 +1,133 @@
+    interaction = float(interaction)
+    chemical_potential = float(chemical_potential)
+
+    # Initialize fermion operator class.
```
Yes, typo! Good catch, and ultimately backs up the code-duplication between the Fermi-Hubbard and Bose-Hubbard functions

josh146 commented on src/openfermion/hamiltonians/_bose_hubbard.py:
```diff
@@ -0,0 +1,133 @@
+                               up_index, down_index)
+
+
+def bose_hubbard(x_dimension, y_dimension, tunneling, interaction,
```
I agree - potentially the bose_hubbard and fermi_hubbard would also differ in their signatures, and the terminology used, so it might make sense to have an underlying _hubbard function, which is then called by a minimal top-level bose_hubbard and fermi_hubbard functions.

josh146 commented on src/openfermion/ops/__init__.py:
```diff
@@ -20,6 +20,10 @@
 from ._fermion_operator import (FermionOperator,
                                 normal_ordered)
+from ._boson_operator import (BosonOperator,
+                                normal_ordered as normal_ordered_boson)
```
I'll modify the pull request to merge the two normal_ordered functions, which branch differently depending on the operator.

josh146 commented:
Thanks for the feedback, it was super fast considering the number of changes!

I'll start fixing the import orderings, and merge the flagged duplicated functionality (namely, bose_hubbard+fermi_hubbard, number operators).

I'm a bit more hesitant to merge and move all the normal_ordered functions into utils (although, in the long run it might make more sense), just because it might affect existing files/tutorials/documentation which include the import

```python
from openfermion.ops import normal_ordered
```

but I'm happy to start the migration if you'd like it contained to the pull request.

babbush commented:
I think in this case it makes sense to move forward with the migration now. We can make a version bump sometime soon. But yes, if you wouldn't mind, please go ahead with this and update the tutorials, etc.

josh146 commented on src/openfermion/hamiltonians/_hubbard_test.py:
```diff
@@ -1,209 +0,0 @@
-#   Licensed under the Apache License, Version 2.0 (the "License");
```
Yes, oops - I'll fix that, and create a single `_hubbard.py` and `_hubbard_test.py` that contains both the Fermi-Hubbard and Bose-Hubbard wrappers, and the underlying hubbard function

josh146 commented:
Done! I've refactored the pull request to take into account the code duplication. I've split the refactor into four commits:

* Merged the number operators
* Merged bose_hubbard and fermi_hubbard so that they share the same underlying function
* Merged all the normal_ordered functions and moved them to utils.

This last commit is quite large, as I had to go and update the source code everywhere that normal_ordered was imported, fix a couple of newly created circular dependencies, and update all the examples and tutorials.

babbush commented:
:+1:
